rcp/org.testeditor.rcp4.views.teststepselection/src/org/testeditor/rcp4/views/teststepselector/TestStepSelector.xtend

```diff
@@ -93,37 +114,47 @@ class TestStepSelector {
 
 	@Inject
 	@Optional
-	def void updateModelAndView(@EventTopic(SELECTOR_UPDATE_MODEL_AND_VIEW) Object data) {
-		logger.debug("updateModelAndView")
-		val amlInjector = amlInjectorProvider.get
-		models = amlInjector.getInstance(AmlModelsProvider).amlModels
-		Display.getDefault.syncExec[internalUpdateView]
+	def void updateView(@EventTopic(SELECTOR_UPDATE_VIEW) Object data) {
+		logger.debug("updateView for " + data)
+		if (data instanceof XtextEditor) {
+			Display.^default.syncExec[updateViewForXtextEditor(data)]
+		}
 	}
 
-	@Inject
-	@Optional
-	def void updateView(@EventTopic(SELECTOR_UPDATE_VIEW) Object data) {
-		logger.debug("updateView")
-		// TODO - check on the project of the editor
-		if (viewer.input === null) {
-			Display.^default.syncExec[internalUpdateView]
+	private def updateViewForXtextEditor(XtextEditor editor) {
+		val projectName = editor.resource.project.name
+
+		var previouslyExpandedElements = viewer.expandedElements.map[toStringPath].toSet
+
+		viewer.input = editor.document.readOnly[readCurrentAMLModel]
+
+		if (currentProject === null || currentProject != projectName) {
+			expandedElementsPerProject.put(currentProject, previouslyExpandedElements)
+			previouslyExpandedElements = expandedElementsPerProject.get(projectName)
+			currentProject = projectName
+		}
+		if (previouslyExpandedElements !== null && !previouslyExpandedElements.empty) {
+			viewer.expandedElements = elementsToExpand(previouslyExpandedElements, viewer.input as Iterable<AmlModel>)
 		}
 	}
 
-	private def internalUpdateView() {
-		// Initial call to the selector
-		if (viewer.input === null) {
-			viewer.input = models
-		} else {
-			// update the view model and conserve the expansion
-			val previouslyExpandedElements = viewer.expandedElements.map[toStringPath].toSet
-			viewer.input = models
-			if(!previouslyExpandedElements.empty){
-				viewer.expandedElements = elementsToExpand(previouslyExpandedElements, models)
+	private def readCurrentAMLModel(XtextResource resource) {
```

Comment:
the method name is misleading. I would expect that it reads the AML model that is included in the passed resource.